### DISCUSSION TITLE
Extract py or mpy from Spike Prime

### DISCUSSION THREAD

**Discussion by QuirkyCort** [pybricks/support#1567] [Q&A]
Are there any ways to extract either the py or mpy file from a Spike Prime? I have a team lose their code, but they still have a copy on the Spike Prime.

I can get into the REPL, but `import os` don't work, so I can't list the files. Tried guessing the file names with `open('XXX', 'r')`, but no success there either.

**Comment by laurensvalk** [comment]
Hi @QuirkyCort - nice to see you here. Great work on your simulator!

The `.py` files are not stored on the hub. As you probably know, the `.mpy` is a compiled format so may still need further deciphering to be useful.

There isn't currently a way to extract the `.mpy` file easily, but since you are quite experienced I can make a small modification to the firmware for you so you can not just read the user data but also your own program.

You'll be able to read the bytes, from which you could piece together the `.mpy` file(s).

Does that sound useful enough?

**Comment by QuirkyCort** [comment]
Thanks. That would help. Just to be sure, writing a new firmware won't erase the user program?

Also, just for my own understanding, doesn't Pybricks use a filesystem? I've worked with micropython on the ESP32 and Pi Pico, and have expected Pybricks to have a filesystem on the internal flash.

**Reply by laurensvalk** [reply]
That is correct. Which firmware is currently on the hub?

> Also, just for my own understanding, doesn't Pybricks use a filesystem?

Also correct. There is an unused megabyte on the secondary storage chip that we use a small section of.

You can still use `import` statements. When you do, the imported program is uploaded also.

We did this so we did not have to touch the existing file system, so people would not lose their original files if they go back.

And also because the Robot Inventor / Spike V2 use a different filesystem than SPIKE V3, so using it or relying on it actually existing is difficult.

We might use the file system in the future, but so far nobody has really needed that much space.

**Reply by QuirkyCort** [reply]
Unfortunately, I do not know the installed Pybricks version, only that the micropython is 1.20. The Spike does not belong to me, and I will only have access to it next Tuesday.

**Reply by laurensvalk** [reply]
That's recent enough :)

**Reply by laurensvalk** [reply]
Do you have a SPIKE Prime Hub to try it out yourself with a simple program first?

- I just tried uploading this program. (On the students hub, of course _skip_ this step)

![image](https://github.com/pybricks/support/assets/12326241/fe986ed0-f6f8-4979-8c26-6926655cc538)

- Make sure you are on the latest firmware.
- Read the existing bytecode chunk by chunk on the REPL. The instructions are in the first post of: https://github.com/pybricks/pybricks-micropython/pull/243

For my test I got:
```
>>> hub.system.storage(PROGRAM_START + 4, read=61)
b'0\x00\x00\x00__main__\x00M\x06\x00\x1f\x03\x01\x10hello.py\x00\x0f\x81w\x05\rHello, world!\x00`\x08\x02\x01\x11\x02#\x004\x01YQc'
```

The part starting with `M\x06` is the `.mpy` file.

**Reply by QuirkyCort** [reply]
Thanks. I don't have a Spike with me now, but I'll try to get one over the weekend and test it out.

**Reply by laurensvalk** [reply]
This functionality is now included in the stable firmware, so you don't need to install anything custom anymore to follow these steps.
